vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java

```diff
@@ -0,0 +1,38 @@
+package br.com.caelum.vraptor.ioc.cdi.extension;
+
+import java.util.Set;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.AfterDeploymentValidation;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.util.AnnotationLiteral;
+
+import br.com.caelum.vraptor.ioc.cdi.EagerInitialized;
+
+/**
+ * When enabled, it allows users to use {@link EagerInitialized} annotation 
+ * to get your application scoped beans initialized just after CDI deployment.
+ *
+ * @author Rodrigo Turini
+ */
+public class EagerApplicationScopedExtension implements Extension {
+	
+	@SuppressWarnings("serial")
+	public void initialize(@Observes AfterDeploymentValidation event, BeanManager beanManager) {
+		
+		Set<Bean<?>> beans = beanManager.getBeans(Object.class, 
+				new AnnotationLiteral<EagerInitialized>() {});
+		
+		for (Bean<?> bean : beans) {
+			Class<?> beanClass = bean.getBeanClass();
+			if(beanClass.isAnnotationPresent(ApplicationScoped.class)){
+				CreationalContext<?> ctx = beanManager.createCreationalContext(bean);
+				beanManager.getReference(bean, beanClass, ctx).toString();
```

Comment:
You got it right. If you use vraptor in a application server, you have all
these possibilites, already. So, I don't see as problem use the same
annotations in other container. Finally, I really believe that our focus
should be the application servers, that is why we are using CDI as our DI
container.

Em qua, 4 de fev de 2015 08:18, Rodrigo Turini <notifications@github.com>
escreveu:

In vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/
> EagerApplicationScopedExtension.java
> <https://github.com/caelum/vraptor4/pull/937#discussion_r24073789>:
>
> > + *
> > + * @author Rodrigo Turini
> > + */
> > +public class EagerApplicationScopedExtension implements Extension {
> > +	
> > +	@SuppressWarnings("serial")
> > +	public void initialize(@Observes AfterDeploymentValidation event, BeanManager beanManager) {
> > +		
> > +		Set<Bean<?>> beans = beanManager.getBeans(Object.class,
> > +				new AnnotationLiteral<EagerInitialized>() {});
> > +		
> > +		for (Bean<?> bean : beans) {
> > +			Class<?> beanClass = bean.getBeanClass();
> > +			if(beanClass.isAnnotationPresent(ApplicationScoped.class)){
> > +				CreationalContext<?> ctx = beanManager.createCreationalContext(bean);
> > +				beanManager.getReference(bean, beanClass, ctx).toString();
>
> @asouza <https://github.com/asouza> I'm not sure if I get your point. The
> idea is just to reuse `@Startup <https://github.com/Startup>` annotation?
> If so, it would be really confusing and frustrating, since we'd add ejb-api
> and all its annotations/interfaces only to reuse a single annotation. It'll
> allow users adding `@DependsOn` trying to order its startup beans, or even
> use `@Stateless <https://github.com/Stateless>` and so on (that will fail
> quietly, since it's not implemented). It'd also allow users to use wrong
> annotations on interceptors, just like `javax.ejb.aroundInvoke`, etc. IMO
> it doesn't worst. If I got the wrong idea, could you elaborate?
> … <#msg-f:1492176150371950013_msg-f:1492170960906541090_>
> Sent from my iPhone
>  On 03/02/2015, at 23:43, Alberto Luiz Souza <notifications@github.com>
> wrote: In vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/
> extension/EagerApplicationScopedExtension.java: > + * > + * @author
> <https://github.com/author> Rodrigo Turini > + */ > +public class
> EagerApplicationScopedExtension implements Extension { > + > +
> @SuppressWarnings <https://github.com/SuppressWarnings>("serial") > +
> public void initialize(@Observes AfterDeploymentValidation event,
> BeanManager beanManager) { > + > + Set<Bean<?>> beans =
> beanManager.getBeans(Object.class, > + new AnnotationLiteral<EagerInitialized>()
> {}); > + > + for (Bean<?> bean : beans) { > + Class<?> beanClass =
> bean.getBeanClass(); > + if(beanClass.isAnnotationPresent(ApplicationScoped.class)){
> > + CreationalContext<?> ctx = beanManager.createCreationalContext(bean);
> > + beanManager.getReference(bean, beanClass, ctx).toString(); I think we
> should use the @Startup <https://github.com/Startup> annotation. It has a
> well defined meaning... Just add the ejb-spec jar and it is done. I don't
> think it is a good practice to create annotations that already exists. Em
> ter, 3 de fev de 2015 20:26, Lucas Cavalcanti <notifications@github.com>
> escreveu: c \ Reply to this email directly or view it on GitHub.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/vraptor4/pull/937/files#r24073789>.
>